source/common/upstream/cluster_manager_impl.cc

```diff
@@ -255,7 +255,10 @@ ClusterManagerImpl::ClusterManagerImpl(
   // the cluster will depend on a non-EDS cluster, so the non-EDS clusters must be loaded first.
   for (const auto& cluster : bootstrap.static_resources().clusters()) {
     // First load all the primary clusters.
-    if (cluster.type() != envoy::api::v2::Cluster::EDS) {
+    if (cluster.type() != envoy::api::v2::Cluster::EDS ||
+        (cluster.type() == envoy::api::v2::Cluster::EDS &&
+         cluster.eds_cluster_config().eds_config().config_source_specifier_case() ==
+             envoy::api::v2::core::ConfigSource::ConfigSourceSpecifierCase::kPath)) {
```

Comment:
I think this makes sense. We should probably generalize and allow any non-xDS source here, including REST. Can you develop this further and write tests, etc? Also, @lambdai, would be interested in your take.

Reply:
I am not sure about REST type because you can provide there cluster name https://www.envoyproxy.io/docs/envoy/latest/configuration/overview/v2_overview#rest-endpoints so there is a chance that cluster won't be defined.

Reply:
I actually think that statement might be stale @lambdai and too strict. So yeah, I think maybe not REST in general.